pkg/controller/validatingadmissionpolicystatus/crd.go

```diff
@@ -0,0 +1,249 @@
+		for gr := range oldSet {
+			if !observedSet.Has(gr) {
+				if set, ok := t.affectedPolicies[gr]; ok {
+					set.Delete(policyName)
+					if set.Len() == 0 {
+						delete(t.affectedPolicies, gr)
+					}
+				}
+			}
+		}
+	}
+	t.policyDependencies[policyName] = observedSet
+	for _, gr := range grs {
+		set, ok := t.affectedPolicies[gr]
+		if !ok {
+			set = sets.New[string]()
+			t.affectedPolicies[gr] = set
+		}
+		set.Insert(policyName)
+	}
+}
+
+// updateDependenciesForPolicyDeletion unregisters the mapping of the policy being deleted.
+func (t *crdTracker) updateDependenciesForPolicyDeletion(policyName string) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	var dependencies []schema.GroupResource
+	if set, ok := t.policyDependencies[policyName]; ok {
+		dependencies = set.UnsortedList()
+	} else {
+		return
+	}
+	delete(t.policyDependencies, policyName)
+	for _, gvk := range dependencies {
+		if set, ok := t.affectedPolicies[gvk]; ok {
+			set.Delete(policyName)
+			if set.Len() == 0 {
+				delete(t.affectedPolicies, gvk)
+			}
+		}
+	}
+}
+
+func CRDVersionToGVK(crd *apiextensionsv1.CustomResourceDefinition, crdVersion *apiextensionsv1.CustomResourceDefinitionVersion) schema.GroupVersionKind {
+	gvk := schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: crdVersion.Name,
+		Kind:    crd.Spec.Names.Kind,
+	}
+	return gvk
+}
+
+type cachedTypeResolver struct {
+	meta.RESTMapper
+	resolver.SchemaResolver
+	crdClient apiextensionsv1client.CustomResourceDefinitionInterface
+
+	cachedCRD    map[schema.GroupResource]*apiextensionsv1.CustomResourceDefinition
+	cachedSchema map[schema.GroupVersionKind]*spec.Schema
+}
+
+func newCachedTypeResolver(restMapper meta.RESTMapper, resolver resolver.SchemaResolver, crdClient apiextensionsv1client.CustomResourceDefinitionInterface) *cachedTypeResolver {
+	return &cachedTypeResolver{
+		RESTMapper:     restMapper,
+		SchemaResolver: resolver,
+		crdClient:      crdClient,
+		cachedCRD:      map[schema.GroupResource]*apiextensionsv1.CustomResourceDefinition{},
+		cachedSchema:   map[schema.GroupVersionKind]*spec.Schema{},
+	}
+}
+
+func (r *cachedTypeResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	if s, ok := r.cachedSchema[gvk]; ok {
+		return s, nil
+	}
+	return r.SchemaResolver.ResolveSchema(gvk)
+}
+
+func (r *cachedTypeResolver) KindsFor(resource schema.GroupVersionResource) ([]schema.GroupVersionKind, error) {
+	if isGroupBuiltin(resource.Group) {
+		return r.RESTMapper.KindsFor(resource)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	crd, ok := r.cachedCRD[resource.GroupResource()]
+	if !ok {
+		var err error
+		// Assuming resource.Resource is plural.
+		// There does not seem to be a good way to reliably to convert the name without the CRD itself
+		crdName := resource.Resource + "." + resource.Group
+		crd, err = r.crdClient.Get(ctx, crdName, metav1.GetOptions{})
```

Comment:
is this doing a live CRD get? why wouldn't we wait for the CRD informer to catch up and requeue the policy?

Reply:
Because (1) the policy can refers to multiple CRDs and the type resolver needs all of them (2) the controller does not store CRDs.

Reply:
The informer stores CRDs, shouldn't we be getting them from the lister, fed by the informer, instead of a live look up?

also, types that aren't built in and aren't backed by CRDs will 404 here perpetually

Reply:
Changed the resolver to use informer.Lister() instead. Thank you.

For types that are neither CRDs or built-ins, unfortunately these are not yet supported as of this PR.